File: knapsack.py
import random

BACKPACK = [["Sleeping bag", 15, 15], ["Rope", 3, 7], ["Switchblade", 2, 10],
            ["Torch", 5, 5], ["Bottle", 9, 8], ["Food", 20, 17]]


def getInitialPopulation():
    population = []
    for i in range(6):
        population.insert(0, (random.randint(0, 1)))
    return population


def getTotalWeight(backpack):
    weight = 0
    for idx, i in enumerate(backpack):
        weight += BACKPACK[idx][1] * i
    return weight


def getTotalReward(backpack):
    reward = 0
    for idx, i in enumerate(backpack):
        reward += BACKPACK[idx][2] * i
    return reward


def weightBackpacks(backpacks):
    return map(getTotalWeight, backpacks)


def rewardsBackpacks(backpacks):
    return map(getTotalReward, backpacks)


def getSelection(selections):
    minimum = 1
    index = random.randint(0, len(selections) - minimum)
    a = selections[index][0]
    selections.pop(index)

    index = random.randint(0, len(selections) - minimum)
    b = selections[index][0]
    selections.pop(index)

    return a, b


def createGenerations():
    lifeCycle = 5
    generations = 50
    populations = []

    for i in range(generations):
        populations.insert(0, (getInitialPopulation()))

    for cycle in range(lifeCycle):
        selections = filter(lambda x: x[1] <= 30, zip(
            populations,
            weightBackpacks(populations),
            rewardsBackpacks(populations)
        ))
        selections = sorted(
            list(selections), key=lambda x: x[2], reverse=True
        )
        selections = selections[:int(generations/2)]
        sub_populations = []

        while len(selections) > 2:
            a, b = getSelection(selections)

            for i in range(int(generations/2)):
                child = []
                for i in range(6):
                    if random.randint(0, 1) == 1:
                        child.insert(0, (a[i]))
                    else:
                        child.insert(0, (b[i]))
                sub_populations.insert(0, (child))
        populations = filter(lambda x: getTotalWeight(x)
                             <= 30, sub_populations)
    return sorted(list(populations), key=lambda x: getTotalReward(x),  reverse=True)


def getBestGeneration():
    generation = (createGenerations())

    return list(
        zip(
            generation,
            weightBackpacks(generation),
            rewardsBackpacks(generation)
        )
    )[0]


def main():
    print("the best items to choose is: {result}".format(
        result=getBestGeneration())
    )


if __name__ == "__main__":
    main()

File: knapsack.go
package knapsack

import (
	"fmt"
	"math/rand"
)

func randomInt(min, max int) int {
	return min + rand.Intn(max-min)
}

func getInitialPopulation() {
	population := []int{}
	for i := 0; i < len(population); i++ {
		population[i] = randomInt(0, 1)
	}
}

func getFirstPopulation(size int) {
	population := []int{}
	for i := 0; i < size; i++ {
			slice[i] = getInitialPopulation()
	}
}

func calculateWeight(backpack []) {
	weight := 0
	for i := 0; i < len(backpack); i++ {
		weight += BACKPACK_DEFAULT[i][1] * i
	}
}

func calculateReward(backpack []) {
	weight := 0
	for i := 0; i < len(backpack); i++ {
		weight += BACKPACK_DEFAULT[i][2] * i
	}
}

func createWeightInBackpacks(backpacks []) {
	return calculateWeight(backpacks)
}

func createRewardsInBackpacks(backpacks []) {
	return calculateReward(backpacks)
}

func createGenerations() {
	lifeCycle := 3
	generations = 50
	populations = getFirstPopulation(generations)

	for i := 0; i < lifeCycle; i++ {
    weights = createWeightBackpacks(populations)
    rewards = createRewardsBackpacks(populations)
    selections = zip(populations, weights, rewards)
    selections = filter(lambda x: x[1] <= 30, selections)
    selections = sorted(list(selections), key=lambda x: x[2],  reverse=True)
    selections = selections[:int(generations/2)]
    population := []int{}

		for i := 2; i < selections; i-- {
				index = random.randint(0, len(selections) - 1)
				a = selections[index][0]
				selections.pop(index)

				index = random.randint(0, len(selections) - 1)
				b = selections[index][0]
				selections.pop(index)

				for i in range(int(size/2)):
					child = []
					for i in range(6):
						if random.randint(0, 1) == 1:
							child.append(a[i])
						else:
							child.append(b[i])
					childs.append(child)
		}
		populations = filter(lambda x: calculateWeight(x) <= 30, childs)
	}
  return sorted(list(populations), key=lambda x: calculateReward(x),  reverse=True)
}
